src/wp-includes/functions.php

```diff
@@ -7830,6 +7830,29 @@ function wp_unique_id( $prefix = '' ) {
 	return $prefix . (string) ++$id_counter;
 }
 
+/**
+ * Generates an incremental ID that is independent per each different prefix.
+ *
+ * It is similar to `wp_unique_id`, but each prefix has its own internal ID
+ * counter to make each prefix independent from each other. The ID starts at 1
+ * and increments on each call. The returned value is not universally unique,
+ * but it is unique across the life of the PHP process and it's stable per
+ * prefix.
+ *
+ * @since 6.4
+ *
+ * @param string $prefix Optional. Prefix for the returned ID. Default empty string.
+ * @return string Incremental ID per prefix.
+ */
+function wp_unique_prefixed_id( $prefix = '' ) {
```

Comment:
Since WP doesn't use typehints yet, I think it might be good to throw a `_doing_it_wrong` or some other warning if a nonstring is passed here since some of them can produce unexpected content.  See: https://3v4l.org/qb1k6

Reply:
@aaronjorbin and I chatted about this in Make/Core slack [starting here in the thread](https://wordpress.slack.com/archives/C02RQBWTW/p1698094191795039?thread_ts=1698066105.240609&cid=C02RQBWTW).

Consensus:

* `string` is expected, non-strings are not expected.
* When receiving a non-string, throw either a `wp_trigger_error()` or `_doing_it_wrong()` and then use the default of an empty string.

Following @azaozz concerns of overusing `_doing_it_wrong()`, i.e. should be used when it will break something and thus must bail out to avoid an error, I think it needs to use `wp_trigger_error()`. 

This change will affect the tests too. I'll prep follow-up commits to this PR to address these changes.

Reply:
Commit 42035dea6f875148d08094e2063d9515f17a61ff adds a defensive guard for a non-string per above -> raises a PHP notice via `wp_trigger_error()` and sets the `$prefix` to an empty string.

@aaronjorbin want to double check the changes?

Reply:
Looks Great. Thank you!